src/irmin/proof_intf.ml

```diff
@@ -180,13 +195,14 @@ module type Env = sig
   (** {2 Modes} *)
 
   val mode : t -> mode option
-  val to_mode : t -> mode -> unit
-  val with_mode : t -> mode -> (unit -> 'a Lwt.t) -> 'a Lwt.t
+  val to_mode : t -> kind -> mode -> unit
+  val with_mode : t -> kind -> mode -> (unit -> 'a Lwt.t) -> 'a Lwt.t
+  val with_stream : stream -> (t -> 'a Lwt.t) -> 'a Lwt.t
```

Comment:
It'd be nice if these 2 functions `with_mode` and `with_stream` had a more consistent signature. Maybe use `with_set` and `with_stream` instead?

Also is there any way to hide the `to_mode` function

Reply:
`to_mode` (not `set_mode`) should be kept for `to_tree`. I'm looking for a better solution for `with_mode`.

Note that `with_mode` is used in `produce_proof`, `verify_proof` and `produce_stream`

Reply:
Upon reflection I decided to go with a more verbose API:

```ml 
  val with_set_produce :
    (t -> start_serialise:(unit -> unit) -> 'a Lwt.t) -> 'a Lwt.t

  val with_set_consume :
    (t -> stop_deserialise:(unit -> unit) -> 'a Lwt.t) -> 'a Lwt.t

  val with_stream_produce :
    (t -> to_stream:(unit -> stream) -> 'a Lwt.t) -> 'a Lwt.t

  val with_stream_consume :
    stream -> (t -> is_empty:(unit -> bool) -> 'a Lwt.t) -> 'a Lwt.t
```

WDYT?

Reply:
looks great :-)